modules/ssl/ssl_engine_pphrase.c

```diff
@@ -600,3 +610,330 @@ int ssl_pphrase_Handle_CB(char *buf, int bufsize, int verify, void *srv)
+    
+    const char *prompt;
+    
+    ppcb_arg = (pphrase_cb_arg_t *) UI_get0_user_data(ui);
+
+    if (ppcb_arg == NULL) {
+        //TODO LOG Properly
+        return 0;
+    }
+
+    sc = mySrvConfig(ppcb_arg->s);
+    
+    if (sc->server->pphrase_dialog_type == SSL_PPTYPE_BUILTIN
+            || sc->server->pphrase_dialog_type == SSL_PPTYPE_PIPE) {
+        prompt = UI_get0_output_string(uis);
+        apr_file_puts(prompt, writetty);
+    }
+
+    return 1;
+}
+
+int passphrase_ui_close(UI *ui)
+{
+    /* Close the pipes if they were opened
+    */
+    if (readtty) {
+        apr_file_close(readtty);
+        apr_file_close(writetty);
+        readtty = writetty = NULL;
+    }
+    return 1;
+}
+
+UI_METHOD *get_passphrase_ui(server_rec *s, apr_pool_t *p, int idx,
+                             const char *pkey_file,
+                             apr_array_header_t **pphrases)
+{
+    UI_METHOD *ui_method;
+
+    ui_method = UI_create_method("Passphrase UI");
+    UI_method_set_opener(ui_method, passphrase_ui_open);
+    UI_method_set_reader(ui_method, passphrase_ui_read);
+    UI_method_set_writer(ui_method, passphrase_ui_write);
+    UI_method_set_closer(ui_method, passphrase_ui_close);
+
+    return ui_method;
+}
+
+void destroy_passphrase_ui_method(UI_METHOD *ui_method)
+{
+    if (ui_method) {
+        UI_destroy_method(ui_method);
+    }
+}
+
+apr_status_t ssl_load_engine_pkey(server_rec *s, apr_pool_t *p, int idx,
+                                  const char *pkey_file,
+                                  apr_array_header_t **pphrases,
+				                  EVP_PKEY **ppkey)
+{
+    SSLModConfigRec *mc = myModConfig(s);
+    SSLSrvConfigRec *sc = mySrvConfig(s);
+    EVP_PKEY *pPrivateKey = NULL;
+    ENGINE *e;
+    pphrase_cb_arg_t ppcb_arg;
+
+    const char *key_id = asn1_table_vhost_key(mc, p, sc->vhost_id, idx);
+
+    UI_METHOD *ui_method;
+
+    ppcb_arg.s                     = s;
+    ppcb_arg.p                     = p;
+    ppcb_arg.aPassPhrase           = *pphrases;
+    ppcb_arg.nPassPhraseCur        = 0;
+    ppcb_arg.cpPassPhraseCur       = NULL;
+    ppcb_arg.nPassPhraseDialog     = 0;
+    ppcb_arg.nPassPhraseDialogCur  = 0;
+    ppcb_arg.bPassPhraseDialogOnce = TRUE;
+    ppcb_arg.key_id                = key_id;
+    ppcb_arg.pkey_file             = pkey_file;
+
+    if (ppkey == NULL) {
+        // LOG the right error
+	    return ssl_die(s);
+    }
+
+    ui_method = get_passphrase_ui(s, p, idx, pkey_file, pphrases);
+
+    /* Try to use an engine, if available, to load the file/URL */
+    if (mc->szCryptoDevice) {
+        if (!(e = ENGINE_by_id(mc->szCryptoDevice))) {
+            ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(01888)
+                         "Init: Failed to load Crypto Device API `%s'",
+                         mc->szCryptoDevice);
+            ssl_log_ssl_error(SSLLOG_MARK, APLOG_EMERG, s);
+            return ssl_die(s);
+        }
+
+        //TODO ADDED for debug
```

Comment:
if useful for debugging maybe enabling it after a specific apache log level would make sense.

Reply:
Yes, I agree. I will find out a suitable log level to enable this.